chia/server/address_manager_sqlite_store.py

```diff
@@ -0,0 +1,119 @@
+import aiosqlite
+
+from chia.server.address_manager import AddressManager, ExtendedPeerInfo, NEW_BUCKETS_PER_ADDRESS
+from pathlib import Path
+from typing import Dict, List, Optional, Tuple
+
+Node = Tuple[int, ExtendedPeerInfo]
+Table = Tuple[int, int]
+
+
+class AddressManagerSQLiteStore:
+    """
+    Supports creation of an AddressManager deserialized from a SQLite database.
+    """
+
+    @classmethod
+    async def create(cls, db_path: Path) -> Optional[AddressManager]:
+        """ """
+        async with aiosqlite.connect(db_path) as connection:
+            await connection.execute("pragma journal_mode=wal")
+            await connection.execute("pragma synchronous=OFF")
+
+            metadata: Dict[str, str] = await cls._get_metadata(connection)
+            address_manager: Optional[AddressManager] = None
+
+            if not await cls._is_empty(metadata):
+                nodes: List[Node] = await cls._get_nodes(connection)
+                new_table_entries: List[Table] = await cls._get_new_table(connection)
+                address_manager = cls._create_address_manager(metadata, nodes, new_table_entries)
+
+            return address_manager
+
+    @classmethod
+    async def _get_metadata(cls, connection: aiosqlite.Connection) -> Dict[str, str]:
+        cursor = await connection.execute("SELECT key, value from peer_metadata")
+        metadata = await cursor.fetchall()
+        await cursor.close()
+        return {key: value for key, value in metadata}
+
+    @classmethod
+    async def _get_nodes(cls, connection: aiosqlite.Connection) -> List[Node]:
+        cursor = await connection.execute("SELECT node_id, value from peer_nodes")
+        nodes_id = await cursor.fetchall()
+        await cursor.close()
+        return [(node_id, ExtendedPeerInfo.from_string(info_str)) for node_id, info_str in nodes_id]
+
+    @classmethod
+    async def _get_new_table(cls, connection: aiosqlite.Connection) -> List[Table]:
+        cursor = await connection.execute("SELECT node_id, bucket from peer_new_table")
+        entries = await cursor.fetchall()
+        await cursor.close()
+        return [(node_id, bucket) for node_id, bucket in entries]
+
+    @classmethod
+    async def _is_empty(cls, metadata: Dict[str, str]) -> bool:
+        if "key" not in metadata:
+            return True
+        if int(metadata.get("new_count", 0)) > 0:
+            return False
+        if int(metadata.get("tried_count", 0)) > 0:
+            return False
+        return True
+
+    @classmethod
+    def _create_address_manager(
+        cls, metadata: Dict[str, str], nodes: List[Node], new_table_entries: List[Table]
+    ) -> AddressManager:
```

Comment:
Doesn't the `AddressManager` have a constructor that lets it be constructed from these dicts and lists?
It seems suspicious that you duplicated the code below, rather than moved it. If it's still in its original location, is there a reason you can't use that original copy of it?

Reply:
I wanted the migration code to be logically identical to what it was before. The reason for duplicating and not reusing/moving is so that changes can be made in AddressManagerStore going forward without affecting migration.